ibm/service/database/resource_ibm_database.go

```diff
@@ -2198,7 +2206,63 @@ func resourceIBMDatabaseInstanceUpdate(context context.Context, d *schema.Resour
 		}
 	}
 
+	if d.HasChange("version") {
+		//TODO LORNA: hook this up to new endpoint
+		version := d.Get("version").(string)
+		skipBackup := d.Get("version_upgrade_skip_backup") == true
+
+		timeoutHelper := TimeoutHelper{Now: time.Now()}
+		expirationDatetime := timeoutHelper.calculateExpirationDatetime(d.Timeout(schema.TimeoutUpdate))
+
+		log.Printf("[INFO] Change version testing %s, %v, %s", version, skipBackup, expirationDatetime)
+
+		// Check if a version upgrade task is already in progress
+		tm := &TaskManager{
+			Client:     cloudDatabasesClient,
+			InstanceID: instanceID,
+		}
+
+		upgradeInProgress, currentTask, err := tm.matchingTaskInProgress("Upgrading instance")
+		if err != nil {
+			return diag.FromErr(fmt.Errorf("[ERROR] Error getting tasks for instance: %w", err))
+		}
+
+		if upgradeInProgress {
+			log.Printf("[INFO] Upgrade task already in progress, waiting for it to complete: %s", *currentTask.ID)
+			_, err = waitForDatabaseTaskComplete(*currentTask.ID, d, meta, d.Timeout(schema.TimeoutUpdate))
+			if err != nil {
+				return diag.FromErr(fmt.Errorf("[ERROR] error waiting for version upgrade task to complete: %w", err))
+			}
+		} else {
```

Comment:
Doing this check on the Terraform client side seems a bit risky. Are we performing a similar check in all other clients, such as the UI, CLI, and API? It might be safer/better to handle this in the backend and fail there, so that all clients don’t need to replicate the same logic.

Reply:
So the API does handle and returns an error. See here https://github.ibm.com/ibm-cloud-databases/middle-end/pull/614#discussion_r10435241

On UI we do check as well but just so we dont display the upgrade button when an upgrade task is running. For CLI we dont do this, it will just return the API error. For terraform, because of timeouts, we thought it a better user experience to return the current upgrade task and to poll until it completes, otherwise they will keep having to run `terraform apply` to see if the task is complete. What risks do you see? And do you think it better we just return the error to terraform users instead?